### DISCUSSION TITLE
How to build custom query?

### DISCUSSION THREAD

**Discussion by winner106** [panter/ra-data-prisma#88] [Q&A] [answered]
hi buddy, I want to build a  custom query for a specify resource? I tried for days but couldn't find a solution……

Could you please give me  some direction? 

Thank you very much!

```js
const AdminApp = () => {
  const dataProvider = useDataProvider({
    client: apolloClient,
    aliasPrefix: "", 
    filters: {}, // custom filters
    queryDialect: "nexus-prisma", // customize query dialect, defaults to nexus-prisma
  });

  if (!dataProvider) {
    return <Loading />;
  }

  return (
    <Admin dataProvider={dataProvider} authProvider={authProvider}>
      <Resource name="Product" {...product} />
      <Resource name="Customer" {...customer} />
      <Resource name="Order" {...order} />
      <Resource name="Activation" {...activation} />
      <Resource name="ActivationWithCustomerProduct" {...activationWithCustomerProduct} />
    </Admin>
  );
};
```

**Comment by macrozone** [comment]
i'll move that to discussion, hope that this is ok.

**Comment by macrozone** [comment]
@winner106 

Not sure what you want to customize, but some hints:

By default it will fetch all fields of a resource for list and detail view.

You can customize which fields are fetched: https://github.com/panter/ra-data-prisma/blob/master/packages/dataprovider/README.md#customize-fetching--virtual-resources

Additionaly, you can define custom filters for lists: https://github.com/panter/ra-data-prisma/blob/master/packages/dataprovider/README.md#custom-filters

**Reply by winner106** [reply]
Thank you very much for this discussion!

By using the backend, the following API `createOneOrderData` is automatically generated, which can be used normally after testing.

![Snipaste_2021-09-01_00-04-56](https://user-images.githubusercontent.com/50519642/131537383-bb68238c-e52d-4dba-9656-ceb464ee998e.png)

I want to use `connectOrCreate` in the frontend. How can I customize the query in dataProvider?

**Reply by macrozone** [reply]
@winner106 If the user would already exist, you could use a ReferenceField (see https://github.com/panter/ra-data-prisma/blob/master/packages/dataprovider/README.md#relations)  (this will use connect)

But manually alter the query is not yet possible. If you have an idea how an api like that could look like, let me know.

@claudiocro i remember that you used to create related records in a project, do you have  a suggestion?

**Comment by winner106** [comment]
I know it's possible to use 'ReferenceField' on the frontend 'connect' to create a new record, but  now  I want to use 'TextInput' for custom input, because I can see that the 'connectOrCreate' API is automatically generated by the backend and it works fine as shown above , it would be a shame if the frontend could not call the  'connectOrCreate'  .  

I have seen some examples in `codesandbox`  where you can customize the query. The url for reference is as follows:
[https://codesandbox.io/s/6vjq0yw49r?file=/src/dataProvider.js](url)
[https://codesandbox.io/s/pkm8051kp0?file=/src/dataProvider/index.js](url)

But I really don't know how to do that in our `ra-data-prisma`  ( set custom parameters and custom query for a resource)

Anyway, thank you very much for your reply. If it is really impossible, I have to give it up regretfully.

**Reply by macrozone** [reply]
there is currently no api to do that,
but it should be doable to add one. How should it look like in your opinion?

**Comment by winner106** [comment]
I can see 'connectOrCreate' in your source code, but I studied it for a long time and didn't know how to modify it. Because my technology is not good enough, I may not be able to give good advice, or you can refer to the 'API' of other React Admin 'Dataproviders' , many thanks!

**Reply by macrozone** [reply]
the question is not how to implement it, but how it should look like in your project.

**Comment by winner106** [comment]
![](https://wx2.sinaimg.cn/large/54eb5d52gy1gu0g6wpincj20rn0po44t.jpg)

**Comment by macrozone** [comment, marked as answer]
@winner106 i proposed a PR: https://github.com/panter/ra-data-prisma/pull/89

see the tests and / or the readme for how to use that. Would that solve your use case?

**Comment by winner106** [comment]
You're such an efficient master, that's great! 
Such api perfectly solved my many days of trouble! `ra-data-prisma` is the most powerful dataprovider! Thank you very much again!

I hope that this need is also helpful to others, my use code is as follows, for reference:
useDataProvider:
```js
  const dataProvider = useDataProvider({
    client: apolloClient,
    aliasPrefix: "", 
    filters: {}, 
    queryDialect: "nexus-prisma", 
    
    customizeInputData: {
      Order: {
        create: (data, params) => ({
          ...data,
          customer: {
            connectOrCreate: {
              create: {
                email: params.email,
              },
              where: {
                email: params.email,
              },
            },
          },
        }),
      },
    }
  });
```

OrderCreate Component
```js
const OrderCreate = (props) => (
  <Create {...props}>
    <SimpleForm>
      <ReferenceInput label="Product Name" source="product" reference="Product">
        <SelectInput optionText="name" />
      </ReferenceInput>
      <TextInput source="email"/>
      <NumberInput source="permitUsers" label="permitUsers" initialValue={3} />
      <DateInput source="orderDate" label="orderDate" initialValue={new Date()} />
    </SimpleForm>
  </Create>
);
```

**Reply by macrozone** [reply]
@winner106 can you try version 6.6.0 ? i did not test it thoroughly and i am happy to get feedback if it works as intended

**Reply by winner106** [reply]
In my project after test, it works perfect, thank you very much!
